Data types/bundle: updated (generate keyword) (#190)
SpinalHDL/SpinalDoc-RTD: source/SpinalHDL/Structuring/parametrization.rst

Elaboration time parameters
------------------------------------------

You can use the whole Scala syntax to provide elaboration time parameters.

The whole syntax means you have the entire power and feature set of the

> Scala language at your disposal to solve parametrization requirements for

Reviewer: ```suggestion
Scala language at your disposal to solve parameterization requirements for
```

your project at the level of complexity you choose.

SpinalHDL does not place any opinionated restrictions on how to achieve
your parametrization goals.  As such there are many Scala design patterns
and a few SpinalHDL helpers that can be used to manage parameters that
are suited to different parameter management scenarios.

Here are some examples and ideas of the possibilities:

 * Hardwired code and constants (not strictly parameter management at all
   but serves to hilight the most basic mechanism, a code change, not
   paramater data change)
 * Constant values provided from a companion object that are static
   constants in Scala.
 * Values provided to Scala class constructor, often a ``case class`` that
   causes Scala to capture those constructor argument values as constants.
 * Regular Scala flow-control syntax, not limited to but including
   conditionals, looping, everything.
 * Config class pattern (examples exist in library items such as

>    UartCtrlConfig, SpiMasterCtrlConfig)

Reviewer: we could link this to e.g. https://spinalhdl.github.io/SpinalDoc-RTD/master/SpinalHDL/Examples/Intermediates%20ones/uart.html#controller-construction-parameters

 * Plugins (examples exist in VexRiscV project to configure the feature set
   the resulting CPU IP core is built with)

>  * Values and information looked from a file or remote based source.

Reviewer: ```suggestion
 * Values and information read from a file or remote based source.
```

Author: Went with `loaded from a file or network based source`

    val bus = MyBus(width)
  }
  

> You can also use global variable defined in Scala object's (companion object

Reviewer: I have commented on that in #193 - let's discuss that there?

    val mySignal = flag generate (Bool())  //Equivalent to "if (flag) in Bool() else null"
  }

> .. note::

Reviewer: Why not make this normal text?

> 

Reviewer: - `generate` is a method/function, not a "keyword like directive" - `flag.generate(Bool())` is exactly the same
- "follows for a value" seemed like a typo to me (reading it a few times I get what you want to say) + this is written quite complicated. How about something along the lines of "generate will evaluate the given expression depending on the predicate, returning `null` if the predicate is false"?

Author: Technically everything is a method/function, no code exists in the JVM that doesn't follow that paradigm of a bytecode entry point.
So the question then is, what would you want to initially teach someone who does not know anything about the subject ?
Can this be done in a way that they grasp the new concept enough to understand it and move to the next thing.
I find the term `keyword like directive` to encapsulate the nature of this one, it looks and acts like the keyword, it is controlling behaviour like a directive, the syntax is expected to be used like a keyword.   The term stays in the domain of describing how to use it, and not what it does.

Agree on the "follows for a value" is not a good explanation, I am trying to emphasis the expression must return a value, not `Unit` / void.

With your wording, it sounds funny to start the explanation about the thing that happens 2nd (evaluation of expression when predicate true).  So starting the explanation with the evaluation of the predicate should be first, which also follows how it is evaluated if there are additional side-effects (other than the value returned) during the optionally evaluated expression when predicate true.  Changing the order/rewording would solve that.

Maybe it can be concisely explained in terms of a `Ternary` operator concepts ? 

The first sentence (or so) is trying to formally define what is does in terms and rules that are easy to work through and difficult to misunderstand.  I don't know if that adds to the perceived complexity of the explanation.

Reviewer: You might find "keyword like directive" to be a good term, but this is not used anywhere else in either Scala or SpinalHDL docs so googling for it you find no helpful search results. Inventing new names will not make this any easier - and might obfuscate what is actually happening.

Back to the wording, let's turn it around then: "If the predicate is true, generate will evaluate the given expression and return the result, otherwise is returns null"?

Author: I like the term it conveys so much to me.

Give this version a go.

I think everything in Scala returns a value, it is the Monadic way, even Unit and null is a value.
But the purpose of referencing it is to make the user think about the value being returned, but maybe that is a given from the code example and expected Scala understanding of the reader.

   The ``generate`` keyword like directive is a mechanism to evaluate the expression
   that follows for a value (in this case the construction of a Bool() type), if the
   previous predicate expression evaluated true (in this case when `flag` is
   true).  If the predicate expression evalutes to false then a null
   value is the result and would be assigned to the Scala val `mySignal`.

>    This maybe used in cases to help parameterize the SpinalHDL hardware

Reviewer: ```suggestion
   This may be used in cases to help parameterize the SpinalHDL hardware
```
